presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -462,6 +506,48 @@ private int compare(Type type, Object first, Object second)
         }
     }
 
+    /**
+     * Replace time zone component of a {@link TimestampWithTimeZoneType} value with a given one, preserving point in time
+     * (equivalent to {@link java.time.ZonedDateTime#withZoneSameInstant}.
+     */
+    private static Object withTimeZone(TimestampWithTimeZoneType type, Object value, TimeZoneKey newZone)
+    {
+        if (type.isShort()) {
+            return packDateTimeWithZone(unpackMillisUtc((long) value), newZone);
+        }
+        LongTimestampWithTimeZone longTimestampWithTimeZone = (LongTimestampWithTimeZone) value;
+        return LongTimestampWithTimeZone.fromEpochMillisAndFraction(longTimestampWithTimeZone.getEpochMillis(), longTimestampWithTimeZone.getPicosOfMilli(), newZone);
+    }
+
+    private static TimeZoneKey getTimeZone(TimestampWithTimeZoneType type, Object value)
+    {
+        if (type.isShort()) {
+            return unpackZoneKey(((long) value));
+        }
+        return TimeZoneKey.getTimeZoneKey(((LongTimestampWithTimeZone) value).getTimeZoneKey());
+    }
+
+    @VisibleForTesting
+    static boolean isTimestampToTimestampWithTimeZoneInjectiveAt(ZoneId zone, Instant instant)
+    {
+        ZoneOffsetTransition transition = zone.getRules().previousTransition(instant);
+        if (transition == null) {
+            return true;
+        }
+        // DST change forward and the instant is ambiguous, being within the 'gap' area non-monotonic remapping
+        return transition.getDuration().isNegative() || !transition.getDateTimeAfter().minusNanos(1).atZone(zone).toInstant().isAfter(instant);
+    }
+
+    private static Instant getInstantWithTruncation(TimestampWithTimeZoneType type, Object value)
```

Comment:
Truncation is correct, because transitions are defined with nanosecond precision (in fact, most likely with second precision only). This is why i truncate here -- all i want is _not_ to round up to the next nanosecond, with could move me out of the ambiguity area.

Reply:
See also "last within" test cases.